projectmanagement/src/test/groovy/life/qbic/projectmanagement/domain/project/experiment/ExperimentalDesignSpec.groovy

```diff
@@ -37,9 +32,160 @@ class ExperimentalDesignSpec extends Specification {
+    }
+
+    def "when a condition is defined with the same label then fail"() {
+        given:
+        def design = new ExperimentalDesign()
+        def variableName = VariableName.create("environment")
+        design.addVariable(variableName.value(), [ExperimentalValue.create("normal",), ExperimentalValue.create("altered")])
+        def controlCondition = Condition.create("control", VariableLevel.create(variableName, ExperimentalValue.create("altered")))
+        design.conditions.add(controlCondition)
+        when: "a condition is defined with identical variable levels"
+        def result = design.defineCondition("control", design.getLevel("environment", ExperimentalValue.create("normal")).get())
+        then: "an exception is thrown"
+        result.isFailure()
+    }
+
+    def "when a condition is not defined in the design, create a new condition"() {
+        given:
+        def design = new ExperimentalDesign()
+        def variableName = VariableName.create("environment")
+        design.addVariable(variableName.value(), [ExperimentalValue.create("normal",), ExperimentalValue.create("altered")])
+        def expectedCondition = Condition.create("control", VariableLevel.create(variableName, ExperimentalValue.create("normal")))
+
+        when: "a condition is defined"
+        def result = design.defineCondition("control", design.getLevel("environment", ExperimentalValue.create("normal")).get())
+
+        then: "the condition is returned"
+        result.isSuccess()
+        result.value().hasSameLevelsDefined(expectedCondition)
+        result.value().equals(expectedCondition)
     }
 
-    //TODO test define condition
     //TODO test add level to variable
+    def "when a level is added to an existing variable then the level is part of the variable"() {
+        given:
+        def design = new ExperimentalDesign()
+
+        def variableName = "environment"
+        def normalValue = ExperimentalValue.create("normal",)
+        def alteredValue = ExperimentalValue.create("altered")
+        def otherValue = ExperimentalValue.create("other")
+
+        design.addVariable(variableName, [normalValue, alteredValue])
+
+        when:
+        def result = design.addLevelToVariable(variableName, otherValue)
+
+        then:
+        result.isSuccess()
+        result.value() == new VariableLevel(VariableName.create(variableName), otherValue)
+        design.variables.get(0).levels().contains(otherValue)
+    }
+
+    def "when a level is added to an non-existent variable then the result is a failure"() {
+        given:
+        def design = new ExperimentalDesign()
+
+        def variableName = "environment"
+        def otherValue = ExperimentalValue.create("other")
+        when:
+        def result = design.addLevelToVariable(variableName, otherValue)
+        then:
+        result.isFailure()
+    }
+
+    def "when a level is added to an existing variable with the level already defined then the result is a success"() {
```

Comment:
why would you allow to have the level duplicated?

Reply:
because it is not :) the request is that the variable has the level. if it already does all is good